packages/driver-azure/src/driver/client.ts

```diff
@@ -0,0 +1,238 @@
+      const vm = await computeClient.virtualMachines.get(resourceGroupName, name)
+      if (vm.provisioningState === 'Deleting') {
+        throw new AzureResourceNotFound(`current vm instance for envId: ${envId} is deleting`)
+      }
+      const addresses = await getIpAddresses(networkClient, vm)
+      return { vm, ...addresses }
+    },
+    createVMInstance: async ({
+      imageRef,
+      sshPublicKey,
+      vmSize,
+      envId,
+    }: {
+      sshPublicKey: string
+      imageRef: ImageReference
+      vmSize: string
+      envId: string
+
+    }): Promise<VMInstance> => {
+      const availableVmSize = await asyncFirst(
+        asyncFilter(size => size.name === vmSize, computeClient.virtualMachineSizes.list(region))
+      )
+      if (!availableVmSize) {
+        throw new Error(`Size ${vmSize} isn't supported for location ${region}`)
+      }
+      const generateId = nameGenerator(envId)
+      const vmName = generateId('vm')
+      const resourceGroupName = generateId('rg')
+      const storageAccountName = generateId('storage')
+      const vnetName = generateId('vet')
+      const subnetName = generateId('subnet')
+      const publicIPName = generateId('pip')
+      const networkInterfaceName = generateId('nic')
+      const ipConfigName = generateId('ipc')
+      const domainNameLabel = generateId('domain')
+      const osDiskName = generateId('osdisk')
+      const nsgName = generateId('nsg')
+      const securityRuleName = generateId('sr')
+      const tags = getTags(profileId, envId)
+
+      void await createResourceGroup(resourceGroupName, region, tags, resourceClient)
+      void await createStorageAccount(
+        storageAccountName,
+        region,
+        resourceGroupName,
+        tags,
+        storageClient
+      )
+      const vnetInfo = await createVnet(
+        tags,
+        vnetName,
+        region,
+        subnetName,
+        resourceGroupName,
+        networkClient
+      )
+      const securityGroup = await createSecurityGroup(
+        tags,
+        vnetName,
+        securityRuleName,
+        resourceGroupName,
+        nsgName,
+        region,
+        networkClient
+      )
+      const publicIPInfo = await createPublicIP(
+        tags,
+        publicIPName,
+        region,
+        resourceGroupName,
+        domainNameLabel,
+        networkClient
+      )
+      const nic = await createNIC(
+        tags,
+        vnetInfo.subnets?.find(s => s.name === subnetName) as Subnet,
+        publicIPInfo,
+        networkInterfaceName,
+        region,
+        ipConfigName,
+        resourceGroupName,
+        securityGroup,
+        networkClient
+      )
+      const vmImageInfo = await findVMImage(region, imageRef, computeClient)
+
+      if (!nic.id || !vmImageInfo?.name || !publicIPInfo.name || !nic.ipConfigurations?.[0].privateIPAddress) {
+        throw new Error('Could not create vm, missing properties')
```

Comment:
maybe you can include some of the checked props in the message so it will be easier to debug

Reply:
sure